File: 567-permutation-in-string/permutation-in-string.py
class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:

        if len(s1)>len(s2):
            return False
        hm = {}
        for i in s1:
            if i not in hm:
                hm[i] = 0
            hm[i] += 1
        
        comp = {}
        for i in range(len(s1)):
            if s2[i] not in comp:
                comp[s2[i]] = 0
            comp[s2[i]] += 1
        
        
        print(comp)
        l = 0
        for i in range(len(s1),len(s2)):
            if comp == hm:
                return True
            comp[s2[l]] -= 1
            if comp[s2[l]] == 0:
                del comp[s2[l]]
            l += 1
            if s2[i] not in comp:
                comp[s2[i]] = 0
            comp[s2[i]] += 1
        
        return comp == hm
            

            

        
        




















        # if len(s1)>len(s2):
        #     return False

        # hs1 = set(s1)

        # hs2 = set(s2[:len(s1)])

        
        # l = 0
        # r = len(s1)
        # while r<len(s2):
        #     if hs1 == hs2:
        #         return True
        #     hs2.remove(s2[l])
        #     hs2.add(s2[r])
        #     l+=1
        #     r+=1
        # return hs1 == hs2

File: 567-permutation-in-string/permutation-in-string.java
class Solution {
    public boolean checkInclusion(String s1, String s2) {
        if(s1.length()>s2.length()){
            return false;
        }

        HashMap<Character, Integer> hm  =  new HashMap<>();
        for(int i=0;i<s1.length();i++){
            if(!hm.containsKey(s1.charAt(i))){
                hm.put(s1.charAt(i), 1);
            }
            else{
                hm.put(s1.charAt(i), hm.get(s1.charAt(i))+1);
            }
        }
        System.out.println(hm);
        //  first window
        HashMap<Character, Integer> hm1  =  new HashMap<>();
        int k = s1.length();
        for(int j=0;j<k;j++){
            if(!hm1.containsKey(s2.charAt(j))){
                hm1.put(s2.charAt(j), 1);
            }
            else{
                hm1.put(s2.charAt(j), hm1.get(s2.charAt(j))+1);
            }
        }
        if(hm.equals(hm1)){
                return true;
            }
        //  remaining windows

        for(int q=k;q<s2.length(); q++){
            System.out.println(hm1);
            
            if(hm1.get(s2.charAt(q-k))>1){
                hm1.put(s2.charAt(q-k),hm1.get(s2.charAt(q-k))-1);
            }
            else{
                hm1.remove(s2.charAt(q-k));
            }
        

            if(!hm1.containsKey(s2.charAt(q))){
                hm1.put(s2.charAt(q), 1);
            }
            else{
                hm1.put(s2.charAt(q), hm1.get(s2.charAt(q))+1);
            }
            System.out.println(hm1);
            if(hm.equals(hm1)){
                return true;
            }

        }



    return false;

    }
}